spacy_curated_transformers/models/listeners.py

```diff
@@ -435,75 +528,244 @@ def __init__(
+        listener: TransformerListenerModelT,
+        frozen: bool = False,
+    ) -> None:
+        """
+        transformer (TransformerModelT):
+            Transformer model to wrap.
+        listener (TransformerListenerModelT):
+            Listener model to wrap.
+        frozen (bool):
+            If the transformer is frozen.
+        """
+        Model.__init__(
+            self,
+            name=self.name,
+            init=wrapped_transformer_and_listener_init,
+            forward=wrapped_transformer_and_listener_forward,
+            dims={"nO": listener.get_dim("nO")},
+            layers=[transformer, listener],
+        )
+
+        # Ensure that the transformer returns the required outputs.
+        transformer.attrs["_all_layer_outputs"] = listener.name in (
+            ScalarWeightingListener.name,
+            TransformerLayersListener.name,
+        )
+
+        # Freeze the embedded transformer if the source pipe was frozen.
+        transformer.attrs["_frozen"] = frozen
+
+        # Remove the marker from the wrapped listener so that it doesn't
+        # get registered to any upstream transformer pipes.
+        assert TransformerListener.SENTINEL in listener.attrs
+        del listener.attrs[TransformerListener.SENTINEL]
+
+        # Ensure that the listener directly reads from the last batch of transformer
+        # outputs stored on it during prediction. If we don't do this, the wrapped
+        # listener will end up using the annotations of the last upstream transformer
+        # pipe that set the  `trf_data` annotation on the Doc object.
+        assert TransformerListener.use_doc_annotations_for_prediction(listener)
+        TransformerListener.set_use_doc_annotations_for_prediction(listener, False)
+
+    @property
+    def frozen_transformer(self) -> bool:
+        return self.layers[0].attrs["_frozen"]
+
+    @frozen_transformer.setter
+    def frozen_transformer(self, value: bool):
+        self.layers[0].attrs["_frozen"] = value
+
+
+def wrapped_transformer_and_listener_init(model: WrappedTransformerAndListener, X, Y):
+    transformer: TransformerModelT = model.layers[0]
+    listener: TransformerListenerModelT = model.layers[1]
+
+    transformer.init(X=X, Y=Y)
+    listener.init(X=X, Y=Y)
+
+
+def wrapped_transformer_and_listener_forward(
+    model: WrappedTransformerAndListener,
+    docs: WrappedTransformerAndListenerInT,
+    is_train: bool,
+) -> Tuple[WrappedTransformerAndListenerOutT, WrappedTransformerAndListenerBackpropT]:
+    transformer: TransformerModelT = model.layers[0]
+    listener: TransformerListenerModelT = model.layers[1]
+    frozen: bool = transformer.attrs["_frozen"]
+    ops = model.ops
+
+    # Follows the same process as `Transformer.update()`.
+    if frozen:
+        outputs = transformer.predict(docs)
+        bp_outputs = None
+        d_outputs = None
+    else:
+        outputs, bp_outputs = transformer(docs, is_train)
+        d_outputs = [
```

Comment:
Do we need want to allocate these arrays when `is_train==False`?

Reply:
No, we can skip that. Fixed.